p2p/starknet/starknet_test.go

```diff
@@ -97,14 +101,102 @@ func TestClientHandler(t *testing.T) {
 	})
 
 	t.Run("get events", func(t *testing.T) {
-		res, cErr := client.RequestEvents(testCtx, &spec.EventsRequest{})
+		eventsPerBlock := [][]*core.Event{
+			{}, // block with no events
+			{
+				{
+					From: randFelt(t),
+					Data: feltSlice(t, 1, randFelt),
+					Keys: feltSlice(t, 1, randFelt),
+				},
+			},
+			{
+				{
+					From: randFelt(t),
+					Data: feltSlice(t, 2, randFelt),
+					Keys: feltSlice(t, 2, randFelt),
+				},
+				{
+					From: randFelt(t),
+					Data: feltSlice(t, 3, randFelt),
+					Keys: feltSlice(t, 3, randFelt),
+				},
+			},
+		}
+		for blockNumber, events := range eventsPerBlock {
+			blockNumber := uint64(blockNumber)
+			mockReader.EXPECT().BlockByNumber(blockNumber).Return(&core.Block{
+				Header: &core.Header{
+					Number: blockNumber,
+				},
+				Receipts: []*core.TransactionReceipt{
+					{Events: events},
+				},
+			}, nil)
+		}
+
+		numOfBlocks := uint64(len(eventsPerBlock))
+		res, cErr := client.RequestEvents(testCtx, &spec.EventsRequest{
+			Iteration: &spec.Iteration{
+				Start: &spec.Iteration_BlockNumber{
+					BlockNumber: 0,
+				},
+				Direction: spec.Iteration_Forward,
+				Limit:     numOfBlocks,
+				Step:      1,
+			},
+		})
 		require.NoError(t, cErr)
 
-		count := uint64(0)
+		var count uint64
 		for evnt, valid := res(); valid; evnt, valid = res() {
+			if count == numOfBlocks {
+				expectedFin := &spec.EventsResponse{
+					Responses: &spec.EventsResponse_Fin{},
+				}
+				assert.True(t, proto.Equal(expectedFin, evnt))
+				count++
+				break
+			}
+
 			assert.Equal(t, count, evnt.Id.Number)
+
+			passedEvents := eventsPerBlock[int(count)]
+			expectedEventsResponse := &spec.EventsResponse_Events{
+				Events: &spec.Events{
+					Items: utils.Map(passedEvents, func(e *core.Event) *spec.Event {
+						adaptFelt := core2p2p.AdaptFelt
+						return &spec.Event{
+							FromAddress: adaptFelt(e.From),
+							Keys:        utils.Map(e.Keys, adaptFelt),
+							Data:        utils.Map(e.Data, adaptFelt),
+						}
+					}),
+				},
+			}
+
+			assert.True(t, proto.Equal(expectedEventsResponse.Events, evnt.Responses.(*spec.EventsResponse_Events).Events))
```

Comment:
Same as above. 

Also, consider using: `event.Responses.GetEvents()` so that type assertion is not required.